recirq/otoc/otoc_circuits.py

```diff
@@ -0,0 +1,571 @@
+                          butterfly_qubit: Union[Any, Sequence[Any]],
+                          cycle: int,
+                          interaction_sequence: Sequence[Set[Tuple[Any, Any]]]
+                          ) -> Tuple[List[Set[Tuple[Any, Any]]],
+                                     List[Set[Tuple[Any, Any]]]]:
+    r"""Compile the qubit pairs that interact in $U$ and $U^\dagger$.
+
+    The lightcones of a measurement qubit and butterfly qubit(s) are
+    calculated based on the order at which the two-qubit gates are applied
+    and number of circuit cycles. Then the pairs of qubits within the
+    lightcones of the measurement qubit and the butterfly qubit(s) for each
+    cycle in $U$ and $U^\dagger$ are returned.
+
+    Args:
+        meas_qubit: The measurement qubit used in the OTOC experiment. Can be
+            specified in any format (e.g. a cirq.GridQubit object or Tuple[
+            int, int]).
+        butterfly_qubit: The qubit(s) to which a perturbation between $U$ and
+            $U^\dagger$ is to be applied.
+        cycle: Number of cycles in $U$.
+        interaction_sequence: The pairs of qubits that interact in each
+        cycle. If len(interaction_sequence) < cycle, the sequence just
+        keeps repeating itself.
+
+    Returns:
+        forward_seq: Each set in the list represents the pairs of qubits that
+            fall within the lightcones of the measurement and butterfly qubits
+            for each cycle in $U$.
+        reverse_seq: Each set in the list represents the pairs of qubits that
+            fall within the lightcones of the measurement and butterfly qubits
+            for each cycle in $U^\dagger$.
+    """
+    num_configs = len(interaction_sequence)
+
+    if np.ndim(meas_qubit) != np.ndim(butterfly_qubit):
+        multiple_butterflies = True
+    else:
+        multiple_butterflies = False
+
+    forward_cone = _extract_light_cone(
+        butterfly_qubit, cycle, interaction_sequence, reverse=False,
+        from_right=True, multiple_butterflies=multiple_butterflies)
+    forward_seq = []
+    for i, cone in enumerate(forward_cone):
+        int_list = []
+        for pair in interaction_sequence[i % num_configs]:
+            if pair[0] in cone or pair[1] in cone:
+                int_list.append(pair)
+        forward_seq.append(set(int_list))
+
+    reverse_cone_0 = _extract_light_cone(
+        meas_qubit, cycle, interaction_sequence,
+        reverse=True, from_right=True, multiple_butterflies=False)
+    reverse_seq = []
+    reverse_seq_0 = list(reversed(forward_seq))
+    for i, int_list in enumerate(reverse_seq_0):
+        new_set = set({})
+        for pair in int_list:
+            if (pair[0] in reverse_cone_0[i]) and (
+                    pair[1] in reverse_cone_0[i]):
+                new_set.add(pair)
+        reverse_seq.append(new_set)
+
+    return forward_seq, reverse_seq
+
+
+def _extract_light_cone(butterfly_qubits: Union[Any, Sequence[Any]],
```

Comment:
To these private functions at the end, can you add a brief (i.e. one line) docstring giving a very concise summary of what they do and/or how they're used by the "main" functions

Reply:
Done.